### ISSUE TITLE
[python3] Build error

### ISSUE THREAD

**Issue by ghlecl** [microsoft/vcpkg#23517]
I am trying to install pybind11, which depends on python.  The build won't work.

**HOST ENVIRONMENT**
- OS: macOS
- Compiler:  Apple clang version 13.0.0 (clang-1300.0.29.3)
                     Target: x86_64-apple-darwin20.6.0
                     Thread model: posix
                     InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

**TO REPRODUCE**
I am in manifest mode.  The manifest is the following:
```json
{
   "$copyright": "Copyright (c) 2022 Ghyslain Leclerc",
   "name": "rs",
   "version-string": "0.1.0",
   "description": "custom python modules",
   "maintainers": [
      "Leclerc, Ghyslain <ghlecl@protonmail.com>"
   ],
   "dependencies": [
      "pybind11",
      "fmt",
      "catch2"
   ]
}
```
`./vcpkg install`

**FAILURE LOGS**
```
 ./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c: In function 'utime_dir_fd':
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c:5165:9: error: 'HAVE_UTIMENSAT_RUNTIME' undeclared (first use in this function); did you mean 'HAVE_FACCESSAT_RUNTIME'?
 5165 |     if (HAVE_UTIMENSAT_RUNTIME) {
      |         ^~~~~~~~~~~~~~~~~~~~~~
      |         HAVE_FACCESSAT_RUNTIME
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c:5165:9: note: each undeclared identifier is reported only once for each function it appears in
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c: In function 'utime_fd':
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c:5201:9: error: 'HAVE_FUTIMENS_RUNTIME' undeclared (first use in this function); did you mean 'HAVE_RENAMEAT_RUNTIME'?
 5201 |     if (HAVE_FUTIMENS_RUNTIME) {
      |         ^~~~~~~~~~~~~~~~~~~~~
      |         HAVE_RENAMEAT_RUNTIME
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c: In function 'utime_nofollow_symlinks':
./../src/v3.10.2-6f3a722579.clean/Modules/posixmodule.c:5242:9: error: 'HAVE_UTIMENSAT_RUNTIME' undeclared (first use in this function); did you mean 'HAVE_FACCESSAT_RUNTIME'?
 5242 |     if (HAVE_UTIMENSAT_RUNTIME) {
```
**CONSOLE OUTPUT** (where ... is just local directories):
```
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):
    Command failed: /usr/bin/make V=1 -j 9 -f Makefile all
    Working Directory: ...buildtrees/python3/x64-osx-dbg/
    See logs for more information:
      ...buildtrees/python3/build-x64-osx-dbg-out.log
      ...buildtrees/python3/build-x64-osx-dbg-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_make.cmake:202 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_install_make.cmake:26 (vcpkg_build_make)
  ports/python3/portfile.cmake:230 (vcpkg_install_make)
  scripts/ports.cmake:145 (include)


Error: Building package python3:x64-osx failed with: BUILD_FAILED
```

**ADDITONNAL CONTEXT**
My vcpkg was up to date (I have pulled and bootstrapped, I even tried the update after, which said there is nothing to do).
[build-x64-osx-dbg-err.log](https://github.com/microsoft/vcpkg/files/8236755/build-x64-osx-dbg-err.log)

**Comment by JackBoosY** [comment]
Can you please also provide log _build-x64-osx-dbg-out.log_?

Thanks.

I assume `HAVE_BUILTIN_AVAILABLE` was not defined.
Related https://github.com/python/cpython/blob/main/Modules/posixmodule.c

**Comment by ghlecl** [comment]
I don't think it was defined.  I did not.

Here is the log.  Thanks
[build-x64-osx-dbg-out.log](https://github.com/microsoft/vcpkg/files/8241533/build-x64-osx-dbg-out.log)
.

**Comment by JackBoosY** [comment]
```cpp
#if defined(__has_builtin)
#if __has_builtin(__builtin_available)
#define HAVE_BUILTIN_AVAILABLE 1
#endif
#endif
```
I think this should be a upstream bug, can you please report this to https://mail.python.org/mailman3/lists/ first?

Related:
https://clang.llvm.org/docs/LanguageExtensions.html#has-builtin
https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available

**Comment by ghlecl** [comment]
Hi.
I have looked at the various mailing lists and could not figure out to which one it belongs.  I then tried to look at the python bug tracker.  Found a related issue.  Did not have time to report yet.  Not sure I'll find the proper channel.  The process seems more involved.
Thanks anyhow for looking into this.

**Comment by ghlecl** [comment]
I have opened this ticket https://bugs.python.org/issue47033 on the python bug tracker.   Seems related to https://bugs.python.org/issue42692.

I have made a few tests.  This code:
```c
#if defined(__APPLE__)

#if defined(__has_builtin)
#if __has_builtin(__builtin_available)
#define HAVE_BUILTIN_AVAILABLE
#else
#undef HAVE_BUILTIN_AVAILABLE
#endif
#endif
#endif

#ifdef HAVE_BUILTIN_AVAILABLE
int val = 1000;
#else
int val = 700;
#endif

#include <stdio.h>

int main() {
   printf( "%d\n", val );
}
```
compiles fine on my computer and has the appropriate behavior.  Then, when compiling python, it fails, whether via vcpkg or via the python sources.  I have added #warnings preprocessor messages and it really seems that the define is not working when compiling the python file, but otherwise works.  I don't know if it's a combination of some versions of OS/Xcode on my machine, or something else.  I have also used Macports to build python from source (forcing the build with -s) and it works.

Anyhow, sorry I am not more knowledgeable to help with my problem.  Let me know if I can do something else.

**Comment by ghlecl** [comment]
OK.  I have had a response on the python bug tracker, which helped me figure out the problem.

Seems there was a conflict with a MacPorts GCC version that I had installed and that was found on my path before the "fake" default GCC of Apple (which points to clang).  I have removed that version from the path and now pybind11 has been installed via vcpkg.

Thanks for you help.  Sorry for taking your time.  I'll close this issue with the comment.
